src/util.rs

```diff
@@ -158,81 +161,79 @@ pub async fn unzip_file(
-
-    if file
-        .extension()
-        .ok_or_else(|| eyre!("Failed to get file extension for {}", file.display()))?
-        == "gz"
-    {
-        dont_spawn_terminal(
-            Command::new(&_7zip_path)
-                .arg("x")
-                .arg(file)
-                .arg(overwrite_arg)
-                .arg(format!("-o{}", tmp_dir.display())),
-        )
-        .status()
-        .await
-        .context("Failed to execute 7zip")?;
-
-        before = list_dir(dest, None).await?.iter().cloned().collect();
-
-        dont_spawn_terminal(
-            Command::new(&_7zip_path)
-                .arg("x")
-                .arg(&tmp_dir)
-                .arg(overwrite_arg)
-                .arg("-ttar")
-                .arg(format!("-o{}", dest.display())),
-        )
-        .status()
-        .await
-        .context("Failed to execute 7zip")?;
-    } else {
-        before = list_dir(dest, None).await?.iter().cloned().collect();
-        dont_spawn_terminal(
-            Command::new(&_7zip_path)
-                .arg("x")
-                .arg(file)
-                .arg(format!("-o{}", dest.display()))
-                .arg(overwrite_arg),
-        )
-        .status()
-        .await
-        .context("Failed to execute 7zip")?;
+
+    let before: HashSet<PathBuf> = list_dir(dest, None).await?.iter().cloned().collect();
+
+    let dest_file_name = dest.file_name().unwrap_or_else(|| std::ffi::OsStr::new("")).to_str().unwrap_or_else(|| "");
+    let temp_dest_dir = tempdir::TempDir::new(dest_file_name).context(format!("Failed to create temporary directory for {}", dest.display()))?;
+    let temp_dest = temp_dest_dir.path();
+
+    if file_extension == "gz" {
+        let tar_gz = std::fs::File::open(file).context(format!("Failed to open file {}", file.display()))?;
+        let tar = GzDecoder::new(tar_gz);
+        let mut archive = Archive::new(tar);
+        archive.set_overwrite(true);
+        archive.unpack(temp_dest).context(format!("Failed to decompress file {}", file.display()))?;
+    } else if file_extension == "zip" {
+        let zip = std::fs::File::open(file).context(format!("Failed to open file {}", file.display()))?;
+        let mut archive = zip::ZipArchive::new(zip).context(format!("Failed to decompress file {}", file.display()))?;
+        archive.extract(temp_dest).context(format!("Failed to decompress file {}", file.display()))?;
+    } else if file_extension == "rar" {
+        let archive = unrar::Archive::new(file.to_str().ok_or_else(|| eyre!("Non-unicode character in file name {}", file.display()))?.to_string());
+        archive
+            .extract_to(temp_dest.to_str().ok_or_else(|| eyre!("Non-unicode character in file name {}", temp_dest.display()))?.to_string())
+            .map_err(|_| eyre!("Failed to decompress file {}", file.display()))?
+            .process()
+            .map_err(|_| eyre!("Failed to decompress file {}", file.display()))?;
```

Comment:
.context

Reply:
note: the following trait bounds were not satisfied:
            `UnrarError<Vec<unrar::archive::Entry>>: eyre::context::ext::StdError`

They didn't make their error type good I guess